File: web/src/lib/athlete-state.ts
import { createClient } from "./supabase/server";

/**
 * UI register gate (0..3) — drives editorial voice on Today and gates pull-quotes.
 * Source of truth: supabase/functions/_shared/athlete-state.ts:computeDataDepth.
 */
export type DataDepth = 0 | 1 | 2 | 3;

/**
 * Fetch the current athlete's data_depth. Returns 0 on any failure or unauth —
 * the safest default (empty/day-zero UI).
 */
export async function getDataDepth(): Promise<DataDepth> {
  const supabase = await createClient();
  const { data, error } = await supabase
    .from("athlete_state")
    .select("data_depth")
    .limit(1)
    .maybeSingle();
  if (error || !data) return 0;
  const v = (data as { data_depth: number | null }).data_depth ?? 0;
  if (v <= 0) return 0;
  if (v === 1) return 1;
  if (v === 2) return 2;
  return 3;
}

/** True when editorial register (italics, pull-quotes, trend deltas) is allowed. */
export function allowsEditorialVoice(depth: DataDepth): boolean {
  return depth >= 2;
}

File: RunningLog/RunningLog/Shared/AthleteState.swift
//
//  AthleteState.swift
//  RunningLog
//
//  Slim iOS readers for athlete_state fields that gate UI behavior.
//

import Foundation
import os
import Supabase

/// data_depth (0..3) — drives editorial register on Today and gates pull-quote surfaces.
/// Source of truth: supabase/functions/_shared/athlete-state.ts:computeDataDepth.
struct AthleteDataDepth {
    let value: Int

    static func fetch() async -> AthleteDataDepth {
        struct Row: Decodable { let data_depth: Int }
        do {
            let rows: [Row] = try await supabase
                .from("athlete_state")
                .select("data_depth")
                .limit(1)
                .execute()
                .value
            return AthleteDataDepth(value: rows.first?.data_depth ?? 0)
        } catch {
            Log.app.error("AthleteDataDepth fetch failed: \(error)")
            return AthleteDataDepth(value: 0)
        }
    }

    var isEmpty: Bool { value == 0 }
    var isMinimal: Bool { value == 1 }
    var isModerate: Bool { value == 2 }
    var isFull: Bool { value == 3 }

    /// True when editorial register (italics, pull-quotes, trend deltas) is allowed.
    var allowsEditorialVoice: Bool { value >= 2 }
}
